### PR TITLE
restart statefulset in case of metadata changes

### PR DESCRIPTION
**Pull request by SherinV**
Porting change for this [issue](https://github.com/open-cluster-management/backlog/issues/13661) into release 2.3

### CODE REVIEW

File: controllers/searchoperator_controller.go [hunk 1]
```diff
@@ -180,14 +181,21 @@ func (r *SearchOperatorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 		return ctrl.Result{}, nil
 	}
 	if persistence {
+		expectedSts := r.expectedStatefulSet(r.Client,
+			instance, true, persistence)
 		//If running PVC deployment nothing to do
-		if persistenceStatus == statusUsingPVC && deployStatus == deploy &&
-			isStatefulSetAvailable(r.Client) && r.isPodRunning(true, 1) {
+		if persistenceStatus == statusUsingPVC && deployStatus == deploy && isStatefulSetAvailable(r.Client) &&
```
Comment (xrajesh) [inline controllers/searchoperator_controller.go:187]: Add a comment here Please - on what all cases we dont need to go anything

File: controllers/searchoperator_controller.go [hunk 2]
```diff
@@ -180,14 +181,21 @@ func (r *SearchOperatorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
...
+			!statefulSetNeedsUpdate(r.Client, expectedSts) && r.isPodRunning(true, 1) {
+			r.Log.Info("Redisgraph Pod running successfully with PVC.")
 			return ctrl.Result{}, nil
 		}
+		expectedSts = r.expectedStatefulSet(r.Client,
+			instance, false, persistence)
 		//If running degraded deployment AND AllowDegradeMode is set
 		if allowdegrade && persistenceStatus == statusDegradedEmptyDir &&
-			deployStatus == deploy && isStatefulSetAvailable(r.Client) && r.isPodRunning(false, 1) {
```
Comment (xrajesh) [inline controllers/searchoperator_controller.go:190]: Here also add a comment please

File: controllers/searchoperator_controller.go [hunk 3]
```diff
@@ -380,150 +393,165 @@ func int32Ptr(i int32) *int32 { return &i }
 
 func int64Ptr(i int64) *int64 { return &i }
 
+func compareLabels(metadataLabels, ssetLabels map[string]string) bool {
```
Comment (xrajesh) [inline controllers/searchoperator_controller.go:396]: Comment for the purpose of the function would be nice .

File: controllers/searchoperator_controller.go [hunk 4]
```diff
@@ -631,32 +660,53 @@ func updateCustomizationCR(kclient client.Client, cr *searchv1alpha1.SearchCusto
...
+			return true
+		} else {
+			log.Info("No changes required for Statefulset")
+			return false
+		}
+	}
+}
+func updateRedisStatefulSet(client client.Client, deployment *appv1.StatefulSet) {
+	createUpdateSTS := statefulSetNeedsUpdate(client, deployment)
+	if !createUpdateSTS {
```
Comment (xrajesh) [inline controllers/searchoperator_controller.go:681]: can we check statefulSetNeedsUpdate - before we call updateRedisStatefulSet ? Now we call update , then come in the function and check if we needs update .

**Review by xrajesh — APPROVED** [review]

File: controllers/searchoperator_controller.go (hunk 4 above)
Comment (SherinV) [reply controllers/searchoperator_controller.go]: Done

File: controllers/searchoperator_controller.go (hunk 3 above)
Comment (SherinV) [reply controllers/searchoperator_controller.go:406]: Done

File: controllers/searchoperator_controller.go (hunk 2 above)
Comment (SherinV) [reply controllers/searchoperator_controller.go:190]: Done

File: controllers/searchoperator_controller.go (hunk 1 above)
Comment (SherinV) [reply controllers/searchoperator_controller.go:190]: Done

**Review by jlpadilla — APPROVED** [review]
/lgtm
/hold

**Pull request merged**